source/distributed/tria.cc

```diff
@@ -3272,6 +3470,11 @@ namespace parallel
           // fix all the flags to make sure we have a consistent mesh
           this->prepare_coarsening_and_refinement ();
 
+          // enforce 2:1 mesh balance over periodic boundaries
+          if (this->smooth_grid & dealii::Triangulation<dim,spacedim>::limit_level_difference_at_vertices)
+            enforce_mesh_balance_over_periodic_boundaries(*this,
```

Comment:
I think this is the wrong place. We have always advertised `prepare_coarsening_and_refinement()` as the function that definitively determines which cells will be coarsened and which will be refined. We need this, for example, for `SolutionTransfer`. 

The right approach is likely to make `prepare_c_and_r()` virtual, and override it in the current class; there, you'd iterate between `::Triangulation::prepare_c_and_r()` and `enforce_mesh_balance_over_boundaries()` until the refine/coarsening flags don't change any more (in much the same way as `::Triangulation::prepare_c_and_r()` already iterates). Can you rewrite the patch in this direction?